### PR TITLE
Generic merge

### PR DESCRIPTION
**Pull request by brycemcanally**
This PR is based on #3836 and contains the implementation of the generic merge (used for merging files and tags). It is basically a refactor of the current merge process with the addition of a generic notion of a stream that implements `next` and `key` functions that are used by the merge for ordering information. Also, the merge process now takes a generic merge function that performs the merging of a set of streams with the same key.

This PR also includes a fuzz test for the merge process which generates files then randomly splits them (both between files and tags) among multiple file sets. These file sets are then merged and compared to the original files to check for the correct file and tag order (as well as the content being there).

### CODE REVIEW

File: src/server/pkg/storage/fileset/merge.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+		}
+		// Convert generic stream to tag stream.
+		tagStream := ss[0].(*tagStream)
+		// Copy tagged data to writer.
+		w.StartTag(tagStream.tags[0].Id)
+		return CopyN(w, tagStream.r, tagStream.tags[0].SizeBytes)
+	}
+}
+
+type mergePQ struct {
```
[the remark this answers is not in the document]
Comment (brycemcanally) [reply src/server/pkg/storage/fileset/merge.go]: Yeah, I agree. I think I was drinking too much of the Go Kool-Aid when writing this. I expanded the identifiers that made sense to me, but my general opinion these days is as the scope of a variable gets smaller, then so should the name.

File: src/server/pkg/storage/fileset/merge.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+		w.StartTag(tagStream.tags[0].Id)
+		return CopyN(w, tagStream.r, tagStream.tags[0].SizeBytes)
+	}
+}
+
+type mergePriorityQueue struct {
+	queue []stream
+	f     mergeFunc
+	size  int
+}
```
Comment (Tryneus) [inline src/server/pkg/storage/fileset/merge.go]: In general, this priority queue implementation seems a little strange - it doesn't encapsulate its properties very well as the `merge` function is very cognizant of the internal workings, and so the code is a bit hard to follow.  Overall not too bad, since its contained within this one file, but it might be nice to at least add a comment block about the invariants of the `mergePriorityQueue`.
